From eedfec03d9f1468baff0ddf636ddfc6500c78d24 Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 15:22:04 -0400
Subject: [infiniband] ocrdma: Avoid reporting wrong completions in case of error CQEs

Message-id: <27dfe268fec1b8e4f9a049b040fbabcd0c2ded51.1411571481.git.dledford@redhat.com>
Patchwork-id: 96619
O-Subject: [Patch RHEL7 29/48] RDMA/ocrdma: Avoid reporting wrong completions in case of error CQEs
Bugzilla: 1081067 1088577
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Selvin Xavier <selvin.xavier@emulex.com>

During cable pull test with a mount over NFS/RDMA, the driver was
reporting error completions when there were no pending requests in the
SQ and RQ.  This was triggering a host crash because of reporting
wrong work req id.  Avoid this crash by adding a check for SQ and RQ
empty condition and prevent reporting completions if queues are empty.

Signed-off-by: Selvin Xavier <selvin.xavier@emulex.com>
Signed-off-by: Devesh Sharma <devesh.sharma@emulex.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit a96ffb1de9d656ce7083277a8badaa1082813498)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/ocrdma/ocrdma_verbs.c b/drivers/infiniband/hw/ocrdma/ocrdma_verbs.c
index 0d7d808..2b68235 100644
--- a/drivers/infiniband/hw/ocrdma/ocrdma_verbs.c
+++ b/drivers/infiniband/hw/ocrdma/ocrdma_verbs.c
@@ -2489,6 +2489,11 @@ static bool ocrdma_poll_err_scqe(struct ocrdma_qp *qp,
    *stop = true;
    expand = false;
   }
+ } else if (is_hw_sq_empty(qp)) {
+  /* Do nothing */
+  expand = false;
+  *polled = false;
+  *stop = false;
  } else {
   *polled = true;
   expand = ocrdma_update_err_scqe(ibwc, cqe, qp, status);
@@ -2594,6 +2599,11 @@ static bool ocrdma_poll_err_rcqe(struct ocrdma_qp *qp, struct ocrdma_cqe *cqe,
    *stop = true;
    expand = false;
   }
+ } else if (is_hw_rq_empty(qp)) {
+  /* Do nothing */
+  expand = false;
+  *polled = false;
+  *stop = false;
  } else {
   *polled = true;
   expand = ocrdma_update_err_rcqe(ibwc, cqe, qp, status);
-- 
1.7.1